Rework the readme

sharadhr opened the pull request:
EDIT for posterity:

Resolves #2463

sharadhr committed:
Initial readme rework

- Remove HTML hyperlinks; markdown generates automatically
- Use backticks \` to demarcate config macros
- Remove `VULKAN_HPP_NO_STD_MODULE` option

sharadhr committed:
Adhere to some markdownlint recommendations

- `-` instead of `*`
- Spaces around lists

asuessenbach commented:
Is there a reason you want to rework the readme?

sharadhr commented:
Hi @asuessenbach! I've been meaning to do this for a while, actually. I have a list of things that I found a little less than ideal with the current readme:

- Be clear on what this repository is actually for (development of the Vulkan-Hpp generator); the actual generated headers can be downloaded from Vulkan-Headers.
- More instructions on installation/usage from the Vulkan SDK, vcpkg, and Conan.
- Rearranging some of the information to flow a bit better; e.g. start with basic `Info` structs, then move on to specific Vulkan types like `Device`, `Image`, etc
  - Some of the information is a bit scattershot; e.g. there are two sections for RAII best practices; these could be combined.
  - Subsume the [RAII programming guide](vk_raii_ProgrammingGuide.md) into the readme, or split the readme up altogether.
  - Also group together the language-side facilities that Vulkan-Hpp uses, e.g. `[[nodiscard]]`, `std::expected`, `std::optional`, etc
- Remove/edit deprecated information, especially pertaining to `vulkan.cppm`
- Miscellaneous grammar, capitalisation, and formatting (e.g. as demonstrated, making all the configuration options demarcated ` ``in backticks`` ` makes it easy to avoid the `\_`stuff in the ToC. 
  - Adhere to some Common Markdown best practices, e.g. spaces around lists, `-` instead of `*`, etc. 
  - Also, the HTML hyperlinks are no longer necessary; as far as I understand, the Common Markdown specification provides a means to automatically generate a table of contents.
  
If you're OK with this, I can proceed with making these changes, naturally subject to your review at the end. I hope to improve the documentation at the end of this exercise.

sharadhr committed:
Introduction and installation

sharadhr committed:
Building section

sharadhr committed:
Minor intro changes

sharadhr committed:
Initial move of stuff into separate markdown files
- Minor changes to main readme so that links work

sharadhr committed:
Changed header levels
- Minor changes in main readme

sharadhr committed:
WIP

sharadhr committed:
Update `Usage.md` for structs
- Add builder pattern

sharadhr committed:
Add one more line break for note

M2-TE commented:
How about adding CMake's `FetchContent` to the "Installation and Usage" section?
Something akin to
```cmake
include(FetchContent)
FetchContent_Declare(vulkan-hpp
    GIT_REPOSITORY "https://github.com/KhronosGroup/Vulkan-Hpp.git"
    GIT_TAG "v1.4.336" # release tag, commit hash or branch name
    GIT_SHALLOW ON)
FetchContent_MakeAvailable(vulkan-hpp)
```

M2-TE commented:
For the "Build Instructions", it is likely best to separate header generation and sample/test building. At the very least, the build section should state that building is purely for those tasks and that headers can be used by just including them.

For merely using the headers (because e.g. they haven't landed in Vulkan-Headers yet), quickly showing how to link against `Vulkan::Hpp` in CMake would be sufficient for those users.

sharadhr committed:
Detail struct usage

sharadhr committed:
Usage rework continues

- Rework flow; start with vulkan fundamentals
- Move things around
- Adhere to 1 sentence, 1 line

sharadhr committed:
Move handles dpcs to `Handles.md`

sharadhr committed:
Add changes from `main`

sharadhr committed:
Rework error-handling section
- Added TODOs for review
- Added section on `std::expected`

sharadhr committed:
Add note about CI status broken

sharadhr committed:
Minor top-level updates

sharadhr committed:
Move some things around in preparation

sharadhr committed:
Minor wording

sharadhr committed:
Move more things around and change header levels

sharadhr committed:
Add headings and TOC for Handles doc

sharadhr committed:
Rewrite and reformat the dispatcher section

sharadhr committed:
Update misc sections;

- Reformat type traits and vkformat traits lists as tables

sharadhr committed:
Update last few sections

sharadhr committed:
Big rewrite for C++ module section

sharadhr commented:
I'm much happier about the Usage document; probably a few final updates left for the last few sections. At this point, I'd like to ask you guys, @M2-TE and @asuessenbach to review what has been done; I have a _lot_ of `TODO`s that you might want to search for and comment on. 

After that, I intend to work on migrating the RAII guide to `Handles.md`, and finally revise the building and installation sections.

M2-TE commented on docs/Building.md:
```diff
@@ -0,0 +1,46 @@
+
+## Minimum requirements
+
+The generator, samples, and tests requires a C++11 compiler. The following compilers are known to work, and tested during releases:
```
The generator actually requires a C++20 compiler: https://github.com/KhronosGroup/Vulkan-Hpp/blob/67ec7c93ba68da2edb2e48f884747bb2db72f15e/CMakeLists.txt#L116

M2-TE commented on docs/Building.md:
```diff
@@ -0,0 +1,46 @@
+
+1. Install dependencies:
+   - Install CMake and git; ensure both are available from a shell.
+   - Install the LunarG Vulkan SDK.
```
The SDK should not be needed for either the generators or the tests/samples. The headers and xml registry are fetched from the `Vulkan-Headers` submodule.
It might be worth noting here that, in order to build samples, you need additional deps on Ubuntu:
https://github.com/KhronosGroup/Vulkan-Hpp/blob/67ec7c93ba68da2edb2e48f884747bb2db72f15e/.github/workflows/build.yml#L79

M2-TE commented on docs/Building.md:
```diff
@@ -0,0 +1,46 @@
+
+### Formatting
+
+If a `clang-format` executable is found by CMake, the define `CLANG_FORMAT_EXECUTABLE` is set accordingly. In that case, the generated headers are formatted using the `.clang-format` file located in the root directory of this project; otherwise, the formatting is left as hard-coded in the generator.
```
likely better to say "variable `CLANG_FORMAT_EXECUTABLE`", rather than "define".

M2-TE replied:
```suggestion
If a `clang-format` executable is found by CMake, the variable `CLANG_FORMAT_EXECUTABLE` is set accordingly. In that case, the generated headers are formatted using the `.clang-format` file located in the root directory of this project; otherwise, the formatting is left as hard-coded in the generator.
```

likely better to say "variable CLANG_FORMAT_EXECUTABLE", rather than "define".

M2-TE commented on docs/Building.md:
````diff
@@ -0,0 +1,46 @@
+    ```git clone --recurse-submodules https://github.com/KhronosGroup/Vulkan-Hpp.git```
+3. Change the shell's working directory to the newly created `Vulkan-Hpp` directory.
+4. Create a build environment with CMake:
+    ```cmake -DVULKAN_HPP_SAMPLES_BUILD=ON -DVULKAN_HPP_SAMPLES_BUILD_WITH_LOCAL_VULKAN_HPP=ON -DVULKAN_HPP_TESTS_BUILD=ON -DVULKAN_HPP_TESTS_BUILD_WITH_LOCAL_VULKAN_HPP=ON -B build```
````
```suggestion
    ```cmake -DVULKAN_HPP_SAMPLES_BUILD=ON -DVULKAN_HPP_TESTS_BUILD=ON -B build```
```

I think you meant VULKAN_HPP_BUILD_WITH_LOCAL_VULKAN_HPP, as it does not have separate versions for tests/samples. Either way, you can probably leave it out as it defaults to ON.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+> This will disable the above-mentioned generated constructors for structs.
+
+C++20 and later versions support [designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization.html#Designated_initializers) for aggregate types.
+Designated initialisers allow for more expressive code: member variable names are mentioned immediately in plain-text, without needing any static analysis.
```
```suggestion
Designated initializers allow for more expressive code: member variable names are mentioned immediately in plain-text, without needing any static analysis.
```

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+                                     .apiVersion         = vk::ApiVersion11 };
+
+// initialize vk::InstanceCreateInfo
+vk::InstanceCreateInfo instanceCreateInfo{ .pApplicationInfo = &applicationInfo };
```
```suggestion
vk::InstanceCreateInfo instanceCreateInfo{ .pApplicationInfo = &applicationInfo };

// initialize vk::InstanceCreateInfo directly within vk::createInstance
instance = vk::createInstance({ .pApplicationInfo = &applicationInfo });
```

This is a style that I have recently started using, which basically inlines your init structs into the respective function calls. It removes the need to have an additional variable name in the scope and, since we know what struct the function expects, is not obfuscated.

For me personally at least, it helped my code readability, especially in the more "busy" sections with lots of init structs. Pretty neat and concise too. Perhaps worth mentioning this style here?

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+<!-- TODO: maybe move this to ./Handles.md? -->
+
+Vulkan-Hpp provides a `vk::` class for each Vulkan handle, such as `vk::Instance`, `vk::Device`, `vk::Buffer`, and so on.
+These classes are binary-compatible with their corresponding C handles; they can be cast back and forth without any overhead.
```
Isn't this true only for 64 bit programs? It might be good to mention and/or link to `VULKAN_HPP_TYPESAFE_CONVERSION` here.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+
+<!-- TODO: this section seems poorly worded and titled. -->
+
+Vulkan-Hpp generates references for pointers to structs. This conversion allows passing temporary structs to functions which can result in shorter code. In case the input is optional and thus accepting a null pointer, the parameter type will be `vk::Optional<T> const&`. This type accepts either a reference to `T` or `nullptr` as input and thus allows optional temporary structs.
```
This part reads in a bit of a confusing way (as you pointed out in TODO as I just noticed). A more concise way of saying these things would be "structs are passed to functions as constant references, therefore allowing temporary structs" or something akin to that. Mentioning stuff like `vk::ArrayProxy` that Vulkan-Hpp provides would probably also fit this category?

For the C++ example below, it might be nice to also see the C++ approach that is similar to C, i.e. making a vk::ImageSubresource struct an passing it, but then showing that you can also create it as a temporary struct.

This could perhaps be combined with the designated initializer stuff as well (or, at least mentioned here), since creating anonymous structs with those is another useful thing to learn about for passing structs.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+   if (result.result != vk::Result::eSuccess)
+   {
+     // handle error code;
+     // physicalDevice.destroyDevice(device, allocator); // cleanup; destroy device
```
This feels a bit out of scope for this example. You could mention cleaning up resources in general instead.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+### Feature and property enumerations
+
+Related to error-handing, Vulkan enumerations require more careful handling of return values.
+The following code is usually typical:
```
```suggestion
The following code is quite common:
```

usually typically normally, when the stars align

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+Adding the third step can potentially result in more efficient code.
+If you intend to use multiple devices, this last step can be omitted, and delegated to the driver instead.
+
+<!-- TODO: what cases? -->
```
I was wondering this as well, I have yet to see a use case for it.

asuessenbach replied:
If I remember correctly, there was some company having their own (private) graphics library as a dll, based on Vulkan-Hpp. And the actual user of that dll doesn't need to know about that.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+- Clang 21.1.8 or later
+- GCC 15.1 or later
+- CMake version 4.2.1 or later
+
```
Here it should link to our build.yaml for an example setup of each platform (currently excluding only gcc).

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+
+# link Vulkan C++ module into user project
+add_executable( YourProject main.cpp Engine.cpp ... )
+target_link_libraries( YourProject PRIVATE VulkanHppModule )
```
```suggestion
# this UUID is still valid as of CMake 4.2.1
# remove this variable if you are setting the UUID via commandline
set( CMAKE_EXPERIMENTAL_CXX_IMPORT_STD d0edc3af-4c50-42ea-a356-e2862fe7a444 )

# UUID had to be set before setting up the project
project( vulkan_hpp_modules_example LANGUAGES CXX )

# find Vulkan SDK (modules were first made available in 1.3.256)
find_package( Vulkan 1.3.256 REQUIRED )

# set up Vulkan C++ module as a static library
add_library( Vulkan-HppModule STATIC )
add_library( Vulkan::HppModule ALIAS Vulkan-HppModule )
target_sources( Vulkan-HppModule PRIVATE
  FILE_SET CXX_MODULES
  TYPE CXX_MODULES
  BASE_DIRS ${Vulkan_INCLUDE_DIR}
  FILES
    ${Vulkan_INCLUDE_DIR}/vulkan/vulkan.cppm
    ${Vulkan_INCLUDE_DIR}/vulkan/vulkan_video.cppm
)

# configure the module target
target_compile_features( Vulkan-HppModule
  PRIVATE cxx_std_23 # (currently) requires C++23 for import std
  PUBLIC cxx_std_20 # can still be consumed as only a C++20 module
)
target_link_libraries( Vulkan-HppModule PUBLIC Vulkan::Vulkan )
set_target_properties( Vulkan-HppModule PROPERTIES
  CXX_SCAN_FOR_MODULES ON
  CXX_MODULE_STD ON
)

# link Vulkan C++ module into user project
add_executable( YourProject main.cpp Engine.cpp ... )
target_link_libraries( YourProject PRIVATE Vulkan::HppModule )
```

M2-TE commented on docs/Usage.md:
````diff
@@ -0,0 +1,966 @@
+
+```cmake
+# Disable exceptions, disable smart handles, disable constructors for structs
+target_compile_definitions( VulkanHppModule PUBLIC
````
```suggestion
target_compile_definitions( Vulkan-HppModule PUBLIC
```
To match the module target name

M2-TE commented on docs/Usage.md:
````diff
@@ -0,0 +1,966 @@
+
+```cmake
+target_link_libraries( VulkanHppModule PUBLIC Vulkan::Headers )
+```
````
```suggestion
```
This was added in the main example, since it is a very common use case.

M2-TE commented on docs/Usage.md:
````diff
@@ -0,0 +1,966 @@
+```
+
+<!-- TODO: this test is gone. @M2-TE? -->
+<!-- An example is provided in [`tests/Cpp20Modules/Cpp20Modules.cpp`](tests/Cpp20Modules/Cpp20Modules.cpp). -->
````
Yup, since most tests now have a module version. This test was previously the only test for modules.

M2-TE commented on docs/Usage.md:
````diff
@@ -0,0 +1,966 @@
+clang++ -std=c++23 <path-to-vulkan-hpp>/vulkan.cppm -precompile -o vulkan.pcm
+clang++ -std=c++23 -fprebuilt-module-path=. main.cpp vulkan.pcm -o main
+./main
+``` -->
````
```suggestion
```
This is indeed very outdated, we do not need to keep this part.

sharadhr replied:
I am thinking of rewriting it to account for `import std` rather than removing it altogether.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+
+## Miscellaneous features and notes
+
+### Extension Inspection
```
It might be worth noting that this very much impacts compilation time.
E.g. Clang in Release mode can take around 2 extra minutes to compile. This is observable in our CI as well.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+| `VULKAN_HPP_CONSTEXPR_20 bool isObsoletedExtension( std::string const & extension )` | `true` if the given extension is obsoleted by some other extension or Vulkan version. |
+| `VULKAN_HPP_CONSTEXPR_20 bool isPromotedExtension( std::string const & extension )` | `true` if the given extension is promoted to some other extension or Vulkan version. |
+
+<!-- TODO: probably need to be updated after recent CI changes. -->
```
Yea this section should be rewritten a little. Samples have additional RAII ones and some relevant flags should be mentioned here (cmake stuff to enable samples, use local headers, precompile headers).

Tests should get the same treatment and show a quick example of how to run them, e.g. via 

```bash
ctest -j --output-on-failure --test-dir build -C Release
```

sharadhr replied:
I was hoping you could provide some input, as the CI changes sort of ran away from me after the first few changes to support modules a few months ago.

M2-TE replied:
Should we just refer to the Build.md section here?

sharadhr replied:
Maybe, but I think at least 1-2 sentences on how it works would be good.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+When you configure your project using CMake, you can enable SAMPLES_BUILD to add some sample projects to your solution. Most of them are ports from the LunarG samples, but there are some more, like CreateDebugUtilsMessenger, InstanceVersion, PhysicalDeviceDisplayProperties, PhysicalDeviceExtensions, PhysicalDeviceFeatures, PhysicalDeviceGroups, PhysicalDeviceMemoryProperties, PhysicalDeviceProper ...
+When you configure your project using CMake, you can enable TESTS_BUILD to add some test projects to your solution. Those tests are just compilation tests and are not required to run.
+
+<!-- TODO: Should we move this to `Configuration.md`? -->
```
I think it's best placed here, rather than in `configuration.md`.
Might be nice to separate these flags better, e.g. group the ones that improve or de..prove? compilation time. Extension inspection should be added here too, with another warning that it very much increases time taken.

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+
+### Handles and functions
+
+<!-- TODO: maybe move this to ./Handles.md? -->
```
It should still have its own section here in usage I believe. A link to ./Handles.md might be good though?

M2-TE commented on docs/Usage.md:
```diff
@@ -0,0 +1,966 @@
+#if ( VULKAN_HPP_DISPATCH_LOADER_DYNAMIC == 1 )
+    // initialize minimal set of function pointers
+    VULKAN_HPP_DEFAULT_DISPATCHER.init();
+#endif
```
```suggestion
VULKAN_HPP_DEFAULT_DISPATCH_LOADER_DYNAMIC_STORAGE

auto main(int argc, char* const argv[]) -> int
{
  // initialize minimal set of function pointers
  VULKAN_HPP_DEFAULT_DISPATCHER.init();
```

Probably best to just use the dynamic dispatcher stuff directly. It's what the example is for, so no need to confuse readers with additional `#ifdef`s.

M2-TE reviewed (commented):
This is a really great PR, fantastic work!
I've added some comments/suggestions where relevant.

> and finally revise the building and installation sections

woops, I also added some comments regarding those sections. If it's too early, just mark them as resolved and I'll revisit them once you have finished the section!

sharadhr commented:
> If it's too early, just mark them as resolved

No worries, I can collapse them and work on your suggestions for the usage doc first. Thanks for being so comprehensive :)

sharadhr commented on docs/Usage.md:
````diff
@@ -0,0 +1,966 @@
+                                     vk::SharingMode::eExclusive, 0, nullptr, vk::ImageLayout::eUndefined});
+```
+
+#### Designated initialisers
````
Sadge. I use British English...

sharadhr committed:
Apply suggestions from @M2-TE

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Resolve comments for `Building.md`

sharadhr committed:
Reword configuration intro

sharadhr committed:
More rework for suggestions; up to error-handling

sharadhr committed:
Suggestions to dynamic loader section implemented

sharadhr committed:
Newlines

sharadhr committed:
Module suggestions by @M2-TE

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Module changes

sharadhr committed:
Apply module dyndis suggestion

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Module doc updates

- Also use `cpp` instead of `c++` for backticks

sharadhr committed:
Update extension inspection table

sharadhr committed:
Initialise more stuff in module example

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Minor change on module names

sharadhr committed:
Write CLI compilation of Vulkan module

sharadhr committed:
Rework Handles

- Move Vk Raii guide; also rename link
- Initial add subheadings for RAII

sharadhr committed:
Minor edit

sharadhr committed:
Fix C++ module CMake

sharadhr committed:
Fix CLI module compilation commands

M2-TE commented on README.md:
```diff
@@ -1,1133 +1,108 @@
+### v1.4.324
+
+With PR [#2226](https://github.com/KhronosGroup/Vulkan-Hpp/pull/2226), the return type of `vk::raii::Device::acquireNextImage2KHR` and `vk::raii::SwapchainKHR::acquireNextImage` has changed from `std::pair<vk::Result,uint32_t>` to the equivalent `vk::ResultValue<uint32_t>`
+
```
```suggestion

### v1.4.329

With PR [#2303](https://github.com/KhronosGroup/Vulkan-Hpp/pull/2303), `import std` was made mandatory when using the named module.

```

Happened a while back, but should still be of interest to some readers

M2-TE commented on README.md:
```diff
@@ -1,1133 +1,108 @@
+### vcpkg
+
+As above, Vulkan-Hpp is provided in the [`vulkan-headers` port](https://vcpkg.io/en/package/vulkan-headers) in [vcpkg](https://github.com/Microsoft/vcpkg).
+Otherwise, you can also install the [`vulkan-sdk-components` port](https://vcpkg.io/en/package/vulkan-sdk-components), which is a vcpkg equivalent to the LunarG-distributed Vulkan SDK.
```
```suggestion
Otherwise, you can also install the [`vulkan-sdk-components`](https://vcpkg.io/en/package/vulkan-sdk-components) port, which is a vcpkg equivalent to the LunarG-distributed Vulkan SDK.
```

M2-TE commented on README.md:
```diff
@@ -1,1133 +1,108 @@
+
+With PR [#2226](https://github.com/KhronosGroup/Vulkan-Hpp/pull/2226), the return type of `vk::raii::Device::acquireNextImage2KHR` and `vk::raii::SwapchainKHR::acquireNextImage` has changed from `std::pair<vk::Result,uint32_t>` to the equivalent `vk::ResultValue<uint32_t>`
+
+## Documentation
```
The documentation section can probably be placed higher. It is quite concise and is often exactly what people are looking for.

Perhaps place it above the `installation and usage`section?

M2-TE commented on docs/Building.md:
```diff
@@ -0,0 +1,53 @@
+
+## Minimum requirements
+
+The generator requires a toolchain that supports C++20; the samples and tests may be compiled with C++11. The following compilers are known to work, and tested during releases:
```
```suggestion
The generator requires a toolchain that supports C++20; the samples and tests may be compiled using at least C++11 with higher versions enabling wider test coverage at up to C++23. The following compilers are known to work, and tested during releases:
```

To let people know that there is a benefit to running tests with higher C++ versions

M2-TE commented on docs/Building.md:
```diff
@@ -0,0 +1,53 @@
+    You may need to specify a generator via `-G`; for a full list of generators execute `cmake -G`.
+
+    To regenerate `vulkan.hpp` from the `vk.xml` XML registry file, add the
+       `-DVULKAN_HPP_RUN_GENERATOR=ON` option to the CMake command line.
```
```suggestion
    To compile samples and tests using freshly generated headers from the `vk.xml` XML registry:
    ```shell
    cmake -B build --preset generator-samples-tests
    ```
```

sharadhr commented on README.md:
```diff
@@ -1,1133 +1,108 @@
+
+- [Overview](#overview)
+- [CI/CD status](#cicd-status)
+- [Installation and usage](#installation-and-usage)
```
This was autogenerated. I can remove it, but I think if I ever edit the readme again, it'll come back...

M2-TE commented:
Leaving a tracking comment here to make sure to wait until v1.4.344 release to test the module example with the ABI breaking changes in #2443.

sharadhr committed:
Keep vcpkg/conan link simple

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Add note for `import std` for `module vulkan`

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Apply build suggestions

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Apply module suggestions

Co-authored-by: Jan Kuhlmann <33833587+M2-TE@users.noreply.github.com>

sharadhr committed:
Move things around in main readme file

sharadhr committed:
Add note about C++ named module and dyndis storage

sharadhr committed:
Add `std::expected` example

sharadhr committed:
Rewrite CMake UUID stuff for CMake 4.3 and later

- Move 3.30 to 4.2 stuff into spoiler

sharadhr committed:
Remove 'compile time issues' altogether, and send warning to configuration file

sharadhr committed:
PascalCase for custom shared handle name

asuessenbach commented on docs/Usage.md:
````diff
@@ -0,0 +1,1006 @@
+  vk::MemoryDedicatedAllocateInfo(image)
+};
+```
+<!-- TODO: need an example here. Also more info and examples: what about the same type, repeated? I don't use structure chains. @asuessenbach? -->
````
Ok, here's a simple example:
```
vk::StructureChain<vk::InstanceCreateInfo, vk::DebugUtilsMessengerCreateInfoEXT, vk::ValidationFlagsEXT> instanceCreateInfoChain = {
  { .pApplicationInfo = &applicationInfo },
  { .messageSeverity = vk::ebugUtilsMessageSeverityFlagBitsEXT::eError },
  { .disabledValidationCheckCount = 1, .pDisabledValidationChecks = &check }
};

if ( dontUseValidationFlags )
{
  instanceCreateInfoChain.unlink<vk::ValidationFlagsEXT>();
}

vk::Instance = vk::createInstance( instanceCreateInfoChain.get<vk::InstanceCreateInfo>() );
```

asuessenbach reviewed (commented):
Indeed, a couple of Todos I didn't see before.
I tried to answer some of them, others (like the one in the Readme.md) just needs to be done.

sharadhr commented:
I think this is something that needs to be discussed with the wider community. I think maybe this should be in its own issue and PR?

There is overlap, but I'm not sure I can speak for everyone about removing `vk::UniqueHandle`.

sharadhr committed:
Address TODOs

sharadhr committed:
Merge remote-tracking branch 'KhronosGroup/main' into rework-readme

sharadhr committed:
Explanation for resultvalue/resultvaluetype

sharadhr committed:
Minor update for default dispatcher

sharadhr committed:
Fix static dispatcher wording

sharadhr committed:
slight reword of Cmake import std uuid

asuessenbach commented:
What a great work!
The readme file is now no longer just a “readme file,” but documentation.

Thanks a lot for this great contribution.
